mods/quick-hide-desktop-icons.wh.cpp

```diff
@@ -145,311 +141,342 @@
+// Double-click detection
 static DWORD g_lastClickTime = 0;
 static POINT g_lastClickPos = {0, 0};
 
-// Find desktop icons window
-HWND FindDesktopIconsWindow() {
-    HWND progman = FindWindowW(L"Progman", nullptr);
-    if (progman) {
-        HWND defView = FindWindowExW(progman, nullptr, L"SHELLDLL_DefView", nullptr);
-        if (defView) {
-            HWND listView = FindWindowExW(defView, nullptr, L"SysListView32", nullptr);
-            if (listView) {
-                Wh_Log(L"Found desktop icons window via Progman: %p", listView);
-                return listView;
-            }
+// Hook
+using DispatchMessageW_t = decltype(&DispatchMessageW);
+static DispatchMessageW_t DispatchMessageW_Original = nullptr;
+
+// Shell process
+static bool g_isShellProcess = false;
+static DWORD g_shellPid = 0;
+static HWND g_lastWorkerW = nullptr;
+
+static DWORD GetShellProcessIdCached() {
+    if (g_shellPid) return g_shellPid;
+    HWND shellWnd = GetShellWindow();
+    if (!shellWnd) return 0;
+    GetWindowThreadProcessId(shellWnd, &g_shellPid);
+    return g_shellPid;
+}
+
+// Convert string to virtual key code
+int StringToVK(PCWSTR str) {
+    if (!str || !*str) return VK_F12;
+    wchar_t tmp[32];
+    wcsncpy_s(tmp, _countof(tmp), str, _TRUNCATE);
+    _wcsupr_s(tmp, _countof(tmp));
+
+    if (tmp[0] == L'F' && wcslen(tmp) >= 2) {
+        int n = _wtoi(tmp + 1);
+        if (n >= 1 && n <= 24) return VK_F1 + (n - 1);
+    }
+    if (wcslen(tmp) == 1) {
+        SHORT vk = VkKeyScanW(tmp[0]);
+        if (vk != -1) return LOBYTE(vk);
+    }
+    if (wcscmp(tmp, L"ESC") == 0) return VK_ESCAPE;
+    if (wcscmp(tmp, L"TAB") == 0) return VK_TAB;
+    if (wcscmp(tmp, L"SPACE") == 0) return VK_SPACE;
+    if (wcscmp(tmp, L"ENTER") == 0 || wcscmp(tmp, L"RETURN") == 0) return VK_RETURN;
+    return VK_F12;
+}
+
+// Clear cached handles
+static void InvalidateCachedDesktopHandles() {
+    g_desktopListView = nullptr;
+    g_defView = nullptr;
+    g_lastWorkerW = nullptr;
+}
+
+// Find desktop DefView
+HWND FindDefView() {
+    if (g_defView && IsWindow(g_defView)) return g_defView;
+
+    DWORD currentPid = GetCurrentProcessId();
+    DWORD shellPid = GetShellProcessIdCached();
+    HWND shellWnd = GetShellWindow();
+    HWND defView = shellWnd ? FindWindowExW(shellWnd, nullptr, L"SHELLDLL_DefView", nullptr) : nullptr;
+
+    if (defView) {
+        DWORD pid;
+        GetWindowThreadProcessId(defView, &pid);
+        if (pid == currentPid || (shellPid && pid == shellPid)) {
+            g_defView = defView;
+            return g_defView;
         }
     }
 
-    HWND workerW = nullptr;
-    while ((workerW = FindWindowExW(nullptr, workerW, L"WorkerW", nullptr))) {
-        HWND defView = FindWindowExW(workerW, nullptr, L"SHELLDLL_DefView", nullptr);
+    HWND worker = g_lastWorkerW && IsWindow(g_lastWorkerW) ? g_lastWorkerW : nullptr;
+    DWORD start = GetTickCount();
+    while ((worker = FindWindowExW(nullptr, worker, L"WorkerW", nullptr))) {
```

Comment:
It'd be better to add a delay in each iteration (e.g. use `Sleep(20);`) to avoid stressing the CPU unnecessarily.